// @appbaseio/reactivemaps/lib/components/basic/GeoDistanceSlider.js
'use strict';

exports.__esModule = true;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _downshift = require('downshift');

var _downshift2 = _interopRequireDefault(_downshift);

var _emotionTheming = require('emotion-theming');

var _actions = require('@appbaseio/reactivecore/lib/actions');

var _helper = require('@appbaseio/reactivecore/lib/utils/helper');

var _Slider = require('rheostat/lib/Slider');

var _Slider2 = _interopRequireDefault(_Slider);

var _types = require('@appbaseio/reactivecore/lib/utils/types');

var _types2 = _interopRequireDefault(_types);

var _Title = require('@appbaseio/reactivesearch/lib/styles/Title');

var _Title2 = _interopRequireDefault(_Title);

var _Input = require('@appbaseio/reactivesearch/lib/styles/Input');

var _Input2 = _interopRequireDefault(_Input);

var _InputIcon = require('@appbaseio/reactivesearch/lib/styles/InputIcon');

var _InputIcon2 = _interopRequireDefault(_InputIcon);

var _SearchSvg = require('@appbaseio/reactivesearch/lib/components/shared/SearchSvg');

var _SearchSvg2 = _interopRequireDefault(_SearchSvg);

var _Slider3 = require('@appbaseio/reactivesearch/lib/styles/Slider');

var _Slider4 = _interopRequireDefault(_Slider3);

var _RangeLabel = require('@appbaseio/reactivesearch/lib/components/range/addons/RangeLabel');

var _RangeLabel2 = _interopRequireDefault(_RangeLabel);

var _SliderHandle = require('@appbaseio/reactivesearch/lib/components/range/addons/SliderHandle');

var _SliderHandle2 = _interopRequireDefault(_SliderHandle);

var _Label = require('@appbaseio/reactivesearch/lib/styles/Label');

var _utils = require('@appbaseio/reactivesearch/lib/utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var GeoDistanceSlider = function (_Component) {
	_inherits(GeoDistanceSlider, _Component);

	function GeoDistanceSlider(props) {
		_classCallCheck(this, GeoDistanceSlider);

		var _this = _possibleConstructorReturn(this, _Component.call(this, props));

		_initialiseProps.call(_this);

		_this.type = 'geo_distance';
		_this.locked = false;
		_this.coordinates = null;
		_this.autocompleteService = null;

		if (props.autoLocation) {
			_this.getUserLocation();
		}
		props.setQueryListener(props.componentId, props.onQueryChange, null);

		props.addComponent(props.componentId);
		_this.setReact(props);

		var currentLocation = null;
		var currentDistance = props.range.start || null;

		if (props.selectedValue) {
			currentLocation = props.selectedValue.location || null;
			currentDistance = props.selectedValue.distance || null;
		} else if (props.value) {
			currentLocation = props.value.location || null;
			currentDistance = props.value.distance || null;
		} else if (props.defaultValue) {
			currentLocation = props.defaultValue.location || null;
			currentDistance = props.defaultValue.distance || null;
		}

		_this.state = {
			currentLocation: currentLocation,
			currentDistance: currentDistance,
			userLocation: null,
			suggestions: [],
			isOpen: false
		};

		_this.getCoordinates(currentLocation, function () {
			if (currentDistance) {
				_this.setDistance(currentDistance);
			}
		});
		return _this;
	}

	GeoDistanceSlider.prototype.componentDidMount = function componentDidMount() {
		this.autocompleteService = new window.google.maps.places.AutocompleteService();
	};

	GeoDistanceSlider.prototype.componentDidUpdate = function componentDidUpdate(prevProps) {
		var _this2 = this;

		(0, _helper.checkPropChange)(this.props.react, prevProps.react, function () {
			return _this2.setReact(prevProps);
		});

		(0, _helper.checkSomePropChange)(this.props, prevProps, ['dataField', 'nestedField'], function () {
			_this2.updateQuery(_this2.state.currentDistance, _this2.props);
		});

		if (this.props.value && !(0, _helper.isEqual)(this.props.value, prevProps.value)) {
			this.setValues(this.props.value);
		} else if (this.props.selectedValue && this.props.selectedValue.distance && this.props.selectedValue.location && !(0, _helper.isEqual)(this.state.currentLocation, this.props.selectedValue.location) && !(0, _helper.isEqual)(this.props.selectedValue, prevProps.selectedValue)) {
			var _props = this.props,
			    onChange = _props.onChange,
			    value = _props.value;

			if (value === undefined) {
				this.setValues(this.props.selectedValue);
			} else if (onChange) {
				onChange(this.props.selectedValue);
			}
		} else if (!(0, _helper.isEqual)(this.props.selectedValue, prevProps.selectedValue) && !this.props.selectedValue) {
			var _props2 = this.props,
			    _value = _props2.value,
			    _onChange = _props2.onChange;

			if (_value === undefined) {
				// eslint-disable-next-line
				this.setState({
					currentLocation: null,
					currentDistance: null
				}, function () {
					_this2.updateQuery(null);
				});
			} else if (_onChange) {
				_onChange({
					location: null,
					distance: null
				});
			}
		}
	};

	GeoDistanceSlider.prototype.componentWillUnmount = function componentWillUnmount() {
		this.props.removeComponent(this.props.componentId);
	};

	GeoDistanceSlider.prototype.setReact = function setReact(props) {
		if (props.react) {
			props.watchComponent(props.componentId, props.react);
		}
	};

	GeoDistanceSlider.prototype.getUserLocation = function getUserLocation() {
		var _this3 = this;

		navigator.geolocation.getCurrentPosition(function (location) {
			var coordinates = location.coords.latitude + ', ' + location.coords.longitude;

			fetch('https://maps.googleapis.com/maps/api/geocode/json?key=' + _this3.props.mapKey + '&v=weekly&latlng=' + coordinates).then(function (res) {
				return res.json();
			}).then(function (res) {
				if (Array.isArray(res.results) && res.results.length) {
					var userLocation = res.results[0].formatted_address;
					_this3.setState({
						userLocation: userLocation
					});
				}
			}).catch(function (e) {
				console.error(e);
			});
		});
	};

	GeoDistanceSlider.prototype.getCoordinates = function getCoordinates(value, cb) {
		var _this4 = this;

		if (value) {
			fetch('https://maps.googleapis.com/maps/api/geocode/json?key=' + this.props.mapKey + '&v=weekly&address=' + value).then(function (res) {
				return res.json();
			}).then(function (res) {
				if (Array.isArray(res.results) && res.results.length) {
					var location = res.results[0].geometry.location;

					_this4.coordinates = location.lat + ', ' + location.lng;
				}
			}).then(function () {
				if (cb) cb();
			}).catch(function (e) {
				console.error(e);
			});
		}
	};

	GeoDistanceSlider.prototype.render = function render() {
		var _this5 = this;

		return _react2.default.createElement(
			_Slider4.default,
			{ primary: true, style: this.props.style, className: this.props.className },
			this.props.title && _react2.default.createElement(
				_Title2.default,
				{ className: (0, _helper.getClassName)(this.props.innerClass, 'title') || null },
				this.props.title
			),
			this.renderSearchBox(),
			_react2.default.createElement(_Slider2.default, {
				min: this.props.range.start,
				max: this.props.range.end,
				values: [this.state.currentDistance],
				onChange: this.handleSlider,
				className: (0, _helper.getClassName)(this.props.innerClass, 'slider'),
				handle: function handle(_ref) {
					var className = _ref.className,
					    style = _ref.style,
					    passProps = _objectWithoutProperties(_ref, ['className', 'style']);

					return _react2.default.createElement(_SliderHandle2.default, _extends({
						style: style,
						className: className
					}, passProps, {
						renderTooltipData: _this5.props.renderTooltipData,
						tooltipTrigger: _this5.props.tooltipTrigger
					}));
				}
			}),
			this.props.rangeLabels ? _react2.default.createElement(
				'div',
				{ className: _Label.rangeLabelsContainer },
				_react2.default.createElement(
					_RangeLabel2.default,
					{
						align: 'left',
						className: (0, _helper.getClassName)(this.props.innerClass, 'label') || null
					},
					this.props.rangeLabels.start
				),
				_react2.default.createElement(
					_RangeLabel2.default,
					{
						align: 'right',
						className: (0, _helper.getClassName)(this.props.innerClass, 'label') || null
					},
					this.props.rangeLabels.end
				)
			) : null
		);
	};

	return GeoDistanceSlider;
}(_react.Component);

var _initialiseProps = function _initialiseProps() {
	var _this6 = this;

	this.setValues = function (selected) {
		_this6.setState({
			currentLocation: selected.location,
			currentDistance: selected.distance
		});
		_this6.getCoordinates(selected.location, function () {
			_this6.setDistance(selected.distance);
		});
	};

	this.defaultQuery = function (coordinates, distance, props) {
		var query = null;
		if (coordinates && distance) {
			var _this6$type, _query;

			query = (_query = {}, _query[_this6.type] = (_this6$type = {
				distance: '' + distance + props.unit
			}, _this6$type[props.dataField] = coordinates, _this6$type), _query);
		}

		if (query && props.nestedField) {
			return {
				query: {
					nested: {
						path: props.nestedField,
						query: query
					}
				}
			};
		}

		return query;
	};

	this.setLocation = function (currentValue) {
		var props = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : _this6.props;

		// ignore state updates when component is locked
		if (props.beforeValueChange && _this6.locked) {
			return;
		}

		_this6.locked = true;

		var performUpdate = function performUpdate() {
			_this6.setState({
				currentLocation: currentValue.value,
				isOpen: false
			}, function () {
				_this6.getCoordinates(currentValue.value, function () {
					if (_this6.state.currentDistance) {
						_this6.updateQuery(_this6.state.currentDistance);
						if (props.onValueChange) {
							props.onValueChange({
								distance: _this6.state.currentDistance,
								location: currentValue.value
							});
						}
					}
					_this6.locked = false;
				});
			});
		};

		(0, _helper.checkValueChange)(props.componentId, { distance: _this6.state.currentDistance, location: currentValue.value }, props.beforeValueChange, performUpdate);
	};

	this.setDistance = function (currentDistance) {
		_this6.setState({
			currentDistance: currentDistance
		}, function () {
			if (_this6.state.currentLocation) {
				_this6.updateQuery(currentDistance, _this6.props);
			}
		});
	};

	this.updateQuery = function (distance) {
		var props = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : _this6.props;
		var componentId = props.componentId,
		    customQuery = props.customQuery,
		    filterLabel = props.filterLabel,
		    showFilter = props.showFilter,
		    URLParams = props.URLParams;

		var value = null;
		if (distance && _this6.state.currentLocation) {
			value = {
				distance: distance,
				location: _this6.state.currentLocation
				// unit: props.unit,
			};
		}
		var query = _this6.defaultQuery(_this6.coordinates, distance, props);
		if (customQuery) {
			var customQueryTobeSet = customQuery(_this6.coordinates, distance, props);
			if (customQueryTobeSet.query) {
				query = customQueryTobeSet.query;
			}
			props.setQueryOptions(_this6.props.componentId, (0, _helper.getOptionsFromQuery)(customQueryTobeSet));
		}
		props.updateQuery({
			componentId: componentId,
			query: query,
			value: value,
			label: filterLabel,
			showFilter: showFilter,
			URLParams: URLParams
		});
	};

	this.renderIcon = function () {
		if (_this6.props.showIcon) {
			return _this6.props.icon || _react2.default.createElement(_SearchSvg2.default, null);
		}
		return null;
	};

	this.onInputChange = function (e) {
		var value = e.target.value;
		var _props3 = _this6.props,
		    valueProp = _props3.value,
		    onChange = _props3.onChange;

		if (valueProp === undefined) {
			_this6.setState({
				currentLocation: value
			});
		} else if (onChange) {
			onChange({ location: value, distance: _this6.state.currentDistance });
		}
		if (value.trim()) {
			if (!_this6.autocompleteService) {
				_this6.autocompleteService = new window.google.maps.places.AutocompleteService();
			}

			var restrictedCountries = _this6.props.countries || [];

			_this6.autocompleteService.getPlacePredictions(_extends({
				input: value,
				componentRestrictions: { country: restrictedCountries }
			}, _this6.props.serviceOptions), function (res) {
				var suggestionsList = res && res.map(function (place) {
					return {
						label: place.description,
						value: place.description
					};
				}) || [];

				_this6.setState({
					suggestions: suggestionsList
				});
			});
		} else {
			_this6.setState({
				suggestions: []
			});
		}
	};

	this.handleFocus = function (event) {
		_this6.setState({
			isOpen: true
		});
		if (_this6.props.onFocus) {
			_this6.props.onFocus(event);
		}
	};

	this.handleOuterClick = function () {
		var _props4 = _this6.props,
		    value = _props4.value,
		    onChange = _props4.onChange;

		if (value === undefined) {
			_this6.setLocation({ value: _this6.state.currentLocation });
		} else if (onChange) {
			onChange({
				location: _this6.state.currentLocation,
				distance: _this6.state.currentDistance
			});
		}
	};

	this.handleStateChange = function (changes) {
		var isOpen = changes.isOpen,
		    type = changes.type;

		if (type === _downshift2.default.stateChangeTypes.mouseUp) {
			_this6.setState({
				isOpen: isOpen
			});
		}
	};

	this.handleLocation = function (data) {
		var _props5 = _this6.props,
		    value = _props5.value,
		    onChange = _props5.onChange;

		if (value === undefined) {
			_this6.setLocation(data);
		} else if (onChange) {
			onChange({ location: data.value, distance: _this6.state.currentDistance });
		}
	};

	this.renderSearchBox = function () {
		var suggestionsList = [].concat(_this6.state.suggestions);
		var _props6 = _this6.props,
		    theme = _props6.theme,
		    themePreset = _props6.themePreset;


		if (_this6.state.userLocation) {
			suggestionsList = [{
				label: 'Use my current location',
				value: _this6.state.userLocation
			}].concat(_this6.state.suggestions);
		}

		return _react2.default.createElement(_downshift2.default, {
			onChange: _this6.handleLocation,
			onOuterClick: _this6.handleOuterClick,
			onStateChange: _this6.handleStateChange,
			isOpen: _this6.state.isOpen,
			itemToString: function itemToString(i) {
				return i;
			},
			render: function render(_ref2) {
				var getInputProps = _ref2.getInputProps,
				    getItemProps = _ref2.getItemProps,
				    isOpen = _ref2.isOpen,
				    highlightedIndex = _ref2.highlightedIndex;
				return _react2.default.createElement(
					'div',
					{ className: _Input.suggestionsContainer },
					_react2.default.createElement(_Input2.default, _extends({
						showIcon: _this6.props.showIcon,
						iconPosition: _this6.props.iconPosition,
						innerRef: _this6.props.innerRef
					}, getInputProps({
						className: (0, _helper.getClassName)(_this6.props.innerClass, 'input'),
						placeholder: _this6.props.placeholder,
						value: _this6.state.currentLocation || '',
						onChange: _this6.onInputChange,
						onBlur: _this6.props.onBlur,
						onFocus: _this6.handleFocus,
						onKeyPress: _this6.props.onKeyPress,
						onKeyDown: _this6.handleKeyDown,
						onKeyUp: _this6.props.onKeyUp
					}), {
						themePreset: themePreset
					})),
					_react2.default.createElement(
						_InputIcon2.default,
						{ iconPosition: _this6.props.iconPosition },
						_this6.renderIcon()
					),
					isOpen && _this6.state.suggestions.length ? _react2.default.createElement(
						'ul',
						{
							className: (0, _Input.suggestions)(themePreset, theme) + ' ' + (0, _helper.getClassName)(_this6.props.innerClass, 'list')
						},
						suggestionsList.slice(0, 11).map(function (item, index) {
							return _react2.default.createElement(
								'li',
								_extends({}, getItemProps({ item: item }), {
									key: item.label,
									style: {
										backgroundColor: highlightedIndex === index ? '#eee' : '#fff'
									}
								}),
								typeof item.label === 'string' ? _react2.default.createElement('div', {
									className: 'trim',
									dangerouslySetInnerHTML: {
										__html: item.label
									}
								}) : item.label
							);
						})
					) : null
				);
			}
		});
	};

	this.handleSlider = function (_ref3) {
		var values = _ref3.values;
		var _props7 = _this6.props,
		    value = _props7.value,
		    onChange = _props7.onChange;

		if (value === undefined) {
			if (values[0] !== _this6.state.currentDistance) {
				_this6.setDistance(values[0]);
			}
		} else if (onChange) {
			// As rheostat do not follow controlled behavior we need to force update the component
			_this6.forceUpdate();
			onChange({ distance: values[0], location: _this6.state.currentLocation });
		}
	};
};

GeoDistanceSlider.propTypes = {
	addComponent: _types2.default.funcRequired,
	mapKey: _types2.default.stringRequired,
	removeComponent: _types2.default.funcRequired,
	selectedValue: _types2.default.selectedValue,
	setQueryListener: _types2.default.funcRequired,
	themePreset: _types2.default.themePreset,
	updateQuery: _types2.default.funcRequired,
	watchComponent: _types2.default.funcRequired,
	// component props
	autoLocation: _types2.default.bool,
	beforeValueChange: _types2.default.func,
	className: _types2.default.string,
	componentId: _types2.default.stringRequired,
	countries: _types2.default.stringArray,
	customQuery: _types2.default.func,
	data: _types2.default.data,
	dataField: _types2.default.stringRequired,
	defaultValue: _types2.default.selectedValue,
	filterLabel: _types2.default.string,
	icon: _types2.default.children,
	iconPosition: _types2.default.iconPosition,
	innerClass: _types2.default.style,
	innerRef: _types2.default.func,
	nestedField: _types2.default.string,
	onBlur: _types2.default.func,
	onChange: _types2.default.func,
	onFocus: _types2.default.func,
	onKeyDown: _types2.default.func,
	onKeyPress: _types2.default.func,
	onKeyUp: _types2.default.func,
	onQueryChange: _types2.default.func,
	onValueChange: _types2.default.func,
	placeholder: _types2.default.string,
	range: _types2.default.range,
	rangeLabels: _types2.default.rangeLabels,
	react: _types2.default.react,
	setQueryOptions: _types2.default.funcRequired,
	showFilter: _types2.default.bool,
	showIcon: _types2.default.bool,
	tooltipTrigger: _types2.default.tooltipTrigger,
	renderTooltipData: _types2.default.func,
	style: _types2.default.style,
	theme: _types2.default.style,
	title: _types2.default.title,
	value: _types2.default.selectedValue,
	unit: _types2.default.string,
	URLParams: _types2.default.bool,
	serviceOptions: _types2.default.props
};

GeoDistanceSlider.defaultProps = {
	className: null,
	placeholder: 'Select a value',
	range: {
		start: 1,
		end: 200
	},
	showFilter: true,
	tooltipTrigger: 'none',
	style: {},
	URLParams: false,
	autoLocation: true,
	unit: 'mi',
	countries: []
};

var mapStateToProps = function mapStateToProps(state, props) {
	return {
		mapKey: state.config.mapKey,
		selectedValue: state.selectedValues[props.componentId] && state.selectedValues[props.componentId].value || null,
		themePreset: state.config.themePreset
	};
};

var mapDispatchtoProps = function mapDispatchtoProps(dispatch) {
	return {
		addComponent: function addComponent(component) {
			return dispatch((0, _actions.addComponent)(component));
		},
		removeComponent: function removeComponent(component) {
			return dispatch((0, _actions.removeComponent)(component));
		},
		updateQuery: function updateQuery(updateQueryObject) {
			return dispatch((0, _actions.updateQuery)(updateQueryObject));
		},
		watchComponent: function watchComponent(component, react) {
			return dispatch((0, _actions.watchComponent)(component, react));
		},
		setQueryListener: function setQueryListener(component, onQueryChange, beforeQueryChange) {
			return dispatch((0, _actions.setQueryListener)(component, onQueryChange, beforeQueryChange));
		},
		setQueryOptions: function setQueryOptions(component, props) {
			return dispatch((0, _actions.setQueryOptions)(component, props));
		}
	};
};

exports.default = (0, _utils.connect)(mapStateToProps, mapDispatchtoProps)((0, _emotionTheming.withTheme)(GeoDistanceSlider));
